src/providers/twitch/TwitchAccount.cpp

```diff
@@ -264,6 +264,25 @@ void TwitchAccount::loadUserstateEmotes(std::weak_ptr<Channel> weakChannel)
             [this, weakChannel](QJsonArray emoteSetArray) {
                 auto emoteData = this->emotes_.access();
                 auto localEmoteData = this->localEmotes_.access();
+
+                std::set<QString> subscriberChannelIDs;
+                for (auto emoteSet : emoteSetArray)
+                {
+                    IvrEmoteSet ivrEmoteSet(emoteSet.toObject());
+                    if (!ivrEmoteSet.tier.isNull())
+                    {
+                        subscriberChannelIDs.insert(ivrEmoteSet.channelId);
+                    }
```

Comment:
This ends up deserializing every emote set twice (once on L271 & once on 290).

Computing all `subscriberChannelIDs` upfront is reasonable if you're concerned about follower emote sets appearing before the corresponding subscriber emote set. In reality, `emote-sets` in `GLOBALUSERSTATE` has the UUIDs are the end, so this is not a concern. (Also the IVR endpoint response has the emote sets in the same order as the request, so misordering wouldn't occur from here either)

Thus, you could move the [L272-L275](https://github.com/Chatterino/chatterino2/blob/6e0c9d611dfbf9d45acb0e94a904c94477a2f16e/src/providers/twitch/TwitchAccount.cpp#L272-L275) logic after [L290](https://github.com/Chatterino/chatterino2/blob/6e0c9d611dfbf9d45acb0e94a904c94477a2f16e/src/providers/twitch/TwitchAccount.cpp#L290) so deserialization only occurs once.

Alternatively, if we're afraid of twitch no longer sorting `emote-sets`, we should consider deserializing once and storing to a vector for reuse

Reply:
Originally I tried storing the `IvrEmoteSet`s in a `std::vector` then sorting it so that the subscriber ones came first. But i was hitting errors about `operator=(&&)` being deleted that I didn't fully understand, so just deserialised it twice instead.

Reply:
When storing to a vector you don't need to sort it (even if we assume twitch doesn't send it already sorted), just iterate twice (as $O(2n) = O(n)$): first iteration does deserialization & populates `subscriberChannelIDs`, second iteration has the remaining logic